modules/src/applications/transfer/coin.rs

```diff
@@ -0,0 +1,157 @@
+use core::str::{from_utf8, FromStr};
+use ibc_proto::cosmos::base::v1beta1::Coin as ProtoCoin;
+use safe_regex::regex;
+use serde::{Deserialize, Serialize};
+
+use super::amount::Amount;
+use super::denom::{BaseDenom, PrefixedDenom};
+use super::error::Error;
+use crate::prelude::*;
+use crate::serializers::serde_string;
+
+/// A `Coin` type with fully qualified `PrefixedDenom`.
+pub type PrefixedCoin = Coin<PrefixedDenom>;
+
+/// A `Coin` type with an unprefixed denomination.
+pub type BaseCoin = Coin<BaseDenom>;
+
+pub type RawCoin = Coin<String>;
+
+/// Coin defines a token with a denomination and an amount.
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Coin<D> {
+    /// Denomination
+    pub denom: D,
+    /// Amount
+    #[serde(with = "serde_string")]
+    pub amount: Amount,
+}
+
+impl<D: FromStr> Coin<D>
+where
+    D::Err: Into<Error>,
+{
+    pub fn from_string_list(coin_str: &str) -> Result<Vec<Self>, Error> {
+        coin_str.split(',').map(FromStr::from_str).collect()
+    }
+}
+
+impl<D: FromStr> FromStr for Coin<D>
+where
+    D::Err: Into<Error>,
+{
+    type Err = Error;
+
+    #[allow(clippy::assign_op_pattern)]
+    fn from_str(coin_str: &str) -> Result<Self, Error> {
+        // Denominations can be 3 ~ 128 characters long and support letters, followed by either
+        // a letter, a number or a separator ('/', ':', '.', '_' or '-').
+        // Loosely copy the regex from here:
+        // https://github.com/cosmos/cosmos-sdk/blob/v0.45.5/types/coin.go#L760-L762
+        let matcher = regex!(br"([0-9]+)([a-zA-Z0-9/:\\._\x2d]+)");
+
+        let (m1, m2) = matcher
+            .match_slices(coin_str.as_bytes())
+            .ok_or_else(|| Error::invalid_coin(coin_str.to_string()))?;
+
+        let amount = from_utf8(m1).map_err(Error::utf8_decode)?.parse()?;
+
+        let denom = from_utf8(m2)
+            .map_err(Error::utf8_decode)?
+            .parse()
+            .map_err(Into::into)?;
+
+        Ok(Coin { amount, denom })
+    }
+}
+
+impl<D: FromStr> TryFrom<ProtoCoin> for Coin<D>
+where
+    D::Err: Into<Error>,
+{
+    type Error = Error;
+
+    fn try_from(proto: ProtoCoin) -> Result<Coin<D>, Self::Error> {
+        let denom = D::from_str(&proto.denom).map_err(Into::into)?;
+        let amount = Amount::from_str(&proto.amount)?;
+        Ok(Self { denom, amount })
+    }
+}
+
+impl<D: ToString> From<Coin<D>> for ProtoCoin {
+    fn from(coin: Coin<D>) -> ProtoCoin {
+        ProtoCoin {
+            denom: coin.denom.to_string(),
+            amount: coin.amount.to_string(),
+        }
+    }
+}
+
+impl From<BaseCoin> for PrefixedCoin {
+    fn from(coin: BaseCoin) -> PrefixedCoin {
+        PrefixedCoin {
+            denom: coin.denom.into(),
+            amount: coin.amount,
+        }
+    }
+}
+
+impl<D: Display> Display for Coin<D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        write!(f, "{}{}", self.amount, self.denom)
```

Comment:
I change the format of coins to follow ibc-go, which has the format `"{amount}{denom}"` instead of `"{amount}-{denom}"`.

Reply:
Previously we only provided a `Display` impl for `PrefixedCoin` because that is being used by the library for logging (in `send_transfer()`). Providing a `Display` impl for all coin types would mean that `PrefixedCoin` would be displayed as `10transfer/channel-0/transfer/channel-1/uatom`. I think I like the old version better. 
An alternative would be to not provide a `Display` impl at all and print the amount and denom separately in logs. Note that a `Coin's` fields are public so users can implement `Display` as they like.

Reply:
Hmm but I try to refactor the code to use the `Coin` type to format output rather than putting `format!("{}{}", amount, denom)` everywhere. Otherwise I guess I'd just define a separate `Token` type inside integration test rather than reusing the `Coin` type.

Reply:
Could you use a `Coin` with a concrete `Denom` and impl `Display` on it in the integration test?

Reply:
Other crates cannot implement traits if they don't own either the type or the trait. In any case, the display format for `PrefixedCoin` also does not match the standard, which requires the path to be hashed.

I wonder if a better way is to use `Debug` for internal use of displaying coins including what type of coin it is, and use `Display` for standard compliance display.

Reply:
I previously meant to suggest you use a newtype to overcome the orphan rule. 
Regarding the display format for `PrefixedCoin`, the ICS doesn't require the trace path to be hashed, it is a Cosmos SDK thing. (I also had this confusion initially. :sweat:)

Reply:
If you like, we can also keep this for now and remove it from the modules after the split as we decided to do with the `IncentivizedPacket` related code earlier.

Reply:
If possible let's keep it for now, since I don't have time to rework #1981 before vacation, but want to keep it in a merge-able state. I can define a separate `Token` type for testing later, which should be easier on one hand, albeit less consistent on the other hand.